mozilla-services/syncstorage-rs
# test: Add Tokenserver integration tests

## Description

Adds a suite of integration tests for Tokenserver. These tests can be run remotely against an arbitrary Tokenserver, meaning they can be run against both the old and the new Tokenserver. All of the tests are passing against the old Tokenserver. Some of the tests were adapted from the tests in [this file](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/tests/test_service.py), but most of them are newly-written.

I also made some changes to the Rust code (there were some errors that the newly-added integration tests caught 🥲)

## Testing

* Add a `MockOAuthVerifier` to [this file](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/verifiers.py) defined like so:
```python
class MockOAuthVerifier(object):
    implements(IOAuthVerifier)
    
    def verify(self, token):
        decoded = jwt.decode(token, options={'verify_signature': False, 'verify_exp': False})
        
        idpclaims = {}
        if decoded.get('fxa-generation') is not None:
            idpclaims['fxa-generation'] = decoded['fxa-generation']
        return {
            'email': decoded.get('sub') + '@' + 'api-accounts.stage.mozaws.net',
            'idpClaims': idpclaims
        }
```
* Start the old Tokenserver with [this configuration file](https://gist.github.com/ethowitz/105744a7d4ffaa74f509b0c0bb3da605)
* In [run.py](https://github.com/mozilla-services/syncstorage-rs/blob/961ab165b7f30ee716fe9699ca04ace23927acac/tools/integration_tests/run.py), comment out [this block of code](https://github.com/mozilla-services/syncstorage-rs/blob/961ab165b7f30ee716fe9699ca04ace23927acac/tools/integration_tests/run.py#L51-L54) and replace it with a call to [this function](https://github.com/mozilla-services/syncstorage-rs/blob/961ab165b7f30ee716fe9699ca04ace23927acac/tools/integration_tests/tokenserver/run.py#L12)
* Ensure that the tests all pass
* Edit the Tokenserver configuration file to comment out line 39 and uncomment lines 40 and 41
* Restart Tokenserver
* Edit [run.py](https://github.com/mozilla-services/syncstorage-rs/blob/961ab165b7f30ee716fe9699ca04ace23927acac/tools/integration_tests/run.py) to replace the call to [run_local_tests()](https://github.com/mozilla-services/syncstorage-rs/blob/961ab165b7f30ee716fe9699ca04ace23927acac/tools/integration_tests/tokenserver/run.py#L12) with a call to [run_end_to_end_tests()](https://github.com/mozilla-services/syncstorage-rs/blob/961ab165b7f30ee716fe9699ca04ace23927acac/tools/integration_tests/tokenserver/run.py#L16)
* Ensure the tests all pass

## Issue(s)

Closes #1048
